### PR TITLE
Fix support for intersections in template literal placeholder types

### LINKED ISSUE #54177: Template Literal Types reduces string union
# Bug Report

When the template literal type is used in a union with strings that match its pattern, the union is reduced:

```ts
type union = 'downcast' | 'dataDowncast' | 'editingDowncast' | `${ string }Downcast`;
//   ^? type union = `${string}Downcast` | "downcast"
```

This is a problem when I want to use it as a parameter type, because then only "downcast" is suggested when calling the function:

```ts
function test( groupName: union ): void {}

test('') // Only "downcast" is suggested
```

As a workaround, I can use function overloading, where one signature has union with all string literals and another only the template literal type, but it would be nice to get this issue fixed:

```ts
function test2( groupName: 'downcast' | 'dataDowncast' | 'editingDowncast' ): void;
function test2( groupName: `${ string }Downcast` ): void;
function test2( groupName: string ): void {}

test2('') // All "downcast", "dataDowncast" and "editingDowncast" are suggested
```

### 🕗 Version & Regression Information

From 4.1 (when template literal types were introduced) to 5.1.0-beta, tested in TypeScript Playground.

### ⏯ Playground Link

[Link to the TypeScript Playground](https://tsplay.dev/wXrnLm)

### 🙁 Actual behavior

Only string literals that don't match the template literal type are displayed.

### 🙂 Expected behavior

All string literals should be shown.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Implements the changes discussed [here](https://github.com/microsoft/TypeScript/issues/54648#issuecomment-1814874112) and fixes #54177 according to our design meeting discussion.

Fixes #54177.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

File: tests/cases/conformance/types/literal/templateLiteralTypesPatterns.ts [hunk 1]
```diff
@@ -202,9 +202,9 @@ export abstract class BB {
 }
 
 // repro from https://github.com/microsoft/TypeScript/issues/54177#issuecomment-1538436654
-function conversionTest(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${string & {}}Downcast`) {}
+function conversionTest(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${string}Downcast` & {}) {}
 conversionTest("testDowncast");
-function conversionTest2(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${{} & string}Downcast`) {}
+function conversionTest2(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | {} & `${string}Downcast`) {}
 conversionTest2("testDowncast");
```
Comment (Andarist) [inline tests/cases/conformance/types/literal/templateLiteralTypesPatterns.ts:208]: I think that those previous test cases are still somewhat valuable. Even if `string & {}` within the placeholder is not meant to enhance open-ended completions those cases are still meant to be type-checked correctly. In TS 5.1 they errored incorrectly: [TS playground](https://www.typescriptlang.org/play?ts=5.1.6#code/GYVwdgxgLglg9mABBBA3ApgJwM7zAFXWygGYAKAc0zhAAcA5AQwFt0AuRAH0QCIATOAHdIjYjy68+jKIwAiQkWIk90fGLDAV5wiKKjjuAAwAkAb2KYYmxADJEpgL4PtiqIYCU9hwCgUYDDh4hMTkPFBEUC66Yu4A3N6gkLAIyGhYuAjBUAAslNR0TKwc3PwK0frKUjJRega8qupWWmW1EiamjraIFk3OLcQeXr5pgZkRuWERNTGxQA)

```suggestion
conversionTest2("testDowncast");
function conversionTest3(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${string & {}}Downcast`) {}
conversionTest3("testDowncast");
function conversionTest4(groupName: | "downcast" | "dataDowncast" | "editingDowncast" | `${{} & string}Downcast`) {}
conversionTest4("testDowncast");
```

Comment (ahejlsberg) [reply tests/cases/conformance/types/literal/templateLiteralTypesPatterns.ts:208]: Ok, I will add them.

**Review by jakebailey — APPROVED** [review]

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by ahejlsberg** [comment]
@jakebailey @DanielRosenwasser So, there's this beauty:

```ts
type Goo<T> = {
  [key: `abc${"def" & { tag: T }}`]: string;
}

type T0 = Goo<void>;  // Index signature key type not instantiated so type parameter leaks
```

By design we don't instantiate index signature key types (that's what mapped types are for), but that in turn means we really can't permit anything generic in tag types. But we have no guaranteed way of detecting whether an object type contains something generic.

Starting to seem like we just want to strip tags in template literal placeholders.

**Comment by ahejlsberg** [comment]
Hmm, the leaking type parameter issue isn't actually anything new:

```ts
type Leak<T> = {
  [key: string & { tag: T }]: string;
}

type T0 = Leak<void>;  // Index signature key type not instantiated so type parameter leaks
```

I seem to recall us discussing this at some point in the past but can't find anything. As I said above, it's harmless, but certainly doesn't look right.

Anyway, if we decide to stay

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-5.3

**Pull request merged**